operator/watchers/node_taint.go

```diff
@@ -0,0 +1,522 @@
+	if err != nil {
+		return false
+	}
+	if len(ciliumPodsInNode) == 0 {
+		return false
+	}
+	for _, ciliumPodInterface := range ciliumPodsInNode {
+		ciliumPod := ciliumPodInterface.(*slim_corev1.Pod)
+		if k8sUtils.GetLatestPodReadiness(ciliumPod.Status) == slim_corev1.ConditionTrue {
+			return true
+		}
+	}
+	return false
+}
+
+// hasAgentNotReadyTaint returns true if the given node has the Cilium Agent
+// Not Ready Node Taint.
+func hasAgentNotReadyTaint(k8sNode *slim_corev1.Node) bool {
+	for _, taint := range k8sNode.Spec.Taints {
+		if taint.Key == ciliumio.AgentNotReadyNodeTaint {
+			return true
+		}
+	}
+	return false
+}
+
+// hostNameIndexFunc index pods by node name.
+func hostNameIndexFunc(obj interface{}) ([]string, error) {
+	switch t := obj.(type) {
+	case *slim_corev1.Pod:
+		return []string{t.Spec.NodeName}, nil
+	}
+	return nil, fmt.Errorf("%w - found %T", errNoPod, obj)
+}
+
+func convertToCiliumPod(obj interface{}) interface{} {
+	switch concreteObj := obj.(type) {
+	case *slim_corev1.Pod:
+		p := &slim_corev1.Pod{
+			TypeMeta: concreteObj.TypeMeta,
+			ObjectMeta: slim_metav1.ObjectMeta{
+				Name:            concreteObj.Name,
+				Namespace:       concreteObj.Namespace,
+				ResourceVersion: concreteObj.ResourceVersion,
+			},
+			Spec: slim_corev1.PodSpec{
+				NodeName: concreteObj.Spec.NodeName,
+			},
+			Status: slim_corev1.PodStatus{
+				Conditions: concreteObj.Status.Conditions,
+			},
+		}
+		*concreteObj = slim_corev1.Pod{}
+		return p
+	case cache.DeletedFinalStateUnknown:
+		pod, ok := concreteObj.Obj.(*slim_corev1.Pod)
+		if !ok {
+			return obj
+		}
+		dfsu := cache.DeletedFinalStateUnknown{
+			Key: concreteObj.Key,
+			Obj: &slim_corev1.Pod{
+				TypeMeta: pod.TypeMeta,
+				ObjectMeta: slim_metav1.ObjectMeta{
+					Name:            pod.Name,
+					Namespace:       pod.Namespace,
+					ResourceVersion: pod.ResourceVersion,
+				},
+				Spec: slim_corev1.PodSpec{
+					NodeName: pod.Spec.NodeName,
+				},
+				Status: slim_corev1.PodStatus{
+					Conditions: pod.Status.Conditions,
+				},
+			},
+		}
+		// Small GC optimization
+		*pod = slim_corev1.Pod{}
+		return dfsu
+	default:
+		return obj
+	}
+}
+
+// setNodeNetworkUnavailableFalse sets Kubernetes NodeNetworkUnavailable to
+// false as Cilium is managing the network connectivity.
+// https://kubernetes.io/docs/concepts/architecture/nodes/#condition
+func setNodeNetworkUnavailableFalse(ctx context.Context, c kubernetes.Interface, nodeGetter slimNodeGetter, nodeName string) error {
+	n, err := nodeGetter.GetK8sSlimNode(nodeName)
+	if err != nil {
+		return err
+	}
+
+	if HasCiliumIsUpCondition(n) {
+		return nil
+	}
+
+	condition := corev1.NodeCondition{
+		Type:               corev1.NodeNetworkUnavailable,
+		Status:             corev1.ConditionFalse,
+		Reason:             ciliumNodeConditionReason,
+		Message:            "Cilium is running on this node",
+		LastTransitionTime: metav1.Now(),
+		LastHeartbeatTime:  metav1.Now(),
```

Comment:
Nit: Maybe store `time.Now()` in a variable and then use it in `LastTransitionTime` and `LastHeartbeatTime` so that the values aren't different?